File: examples/fuel.rs
//! Example of limiting a WebAssembly function's runtime using "fuel consumption".

// You can execute this example with `cargo run --example fuel`

use wasmtime::*;

fn main() -> Result<()> {
    let mut config = Config::new();
    config.consume_fuel(true);
    let engine = Engine::new(&config)?;
    let mut store = Store::new(&engine, ());
    store.set_fuel(10_000)?;
    let module = Module::from_file(store.engine(), "examples/fuel.wat")?;
    let instance = Instance::new(&mut store, &module, &[])?;

    // Invoke `fibonacci` export with higher and higher numbers until we exhaust our fuel.
    let fibonacci = instance.get_typed_func::<i32, i32>(&mut store, "fibonacci")?;
    for n in 1.. {
        let fuel_before = store.get_fuel().unwrap();
        let output = match fibonacci.call(&mut store, n) {
            Ok(v) => v,
            Err(e) => {
                assert_eq!(e.downcast::<Trap>()?, Trap::OutOfFuel);
                println!("Exhausted fuel computing fib({n})");
                break;
            }
        };
        let fuel_consumed = fuel_before - store.get_fuel().unwrap();
        println!("fib({n}) = {output} [consumed {fuel_consumed} fuel]");
        store.set_fuel(10_000)?;
    }
    Ok(())
}

File: examples/fuel.cc
/*
Example of instantiating of the WebAssembly module and invoking its exported
function.

You can compile and run this example on Linux with:

   cargo build --release -p wasmtime-c-api
   c++ examples/fuel.cc -std=c++20 \
       -I crates/c-api/include \
       -I crates/c-api/wasm-c-api/include \
       target/release/libwasmtime.a \
       -lpthread -ldl -lm \
       -o fuel
   ./fuel

Note that on Windows and macOS the command will be similar, but you'll need
to tweak the `-lpthread` and such annotations.
*/

#include <fstream>
#include <iostream>
#include <sstream>
#include <wasmtime.hh>

using namespace wasmtime;

std::string readFile(const char *name) {
  std::ifstream watFile;
  watFile.open(name);
  std::stringstream strStream;
  strStream << watFile.rdbuf();
  return strStream.str();
}

const size_t kStoreFuel = 10000;

int main() {
  Config config;
  config.consume_fuel(true);
  Engine engine(std::move(config));
  Store store(engine);
  store.context().set_fuel(kStoreFuel).unwrap();

  auto wat = readFile("examples/fuel.wat");
  Module module = Module::compile(engine, wat).unwrap();
  Instance instance = Instance::create(store, module, {}).unwrap();
  Func fib = std::get<Func>(*instance.get(store, "fibonacci"));

  // Call it repeatedly until it fails
  for (int32_t n = 1;; n++) {
    auto result = fib.call(store, {n});
    if (!result) {
      std::cout << "Exhausted fuel computing fib(" << n << ")\n";
      break;
    }
    uint64_t consumed = kStoreFuel - store.context().get_fuel().unwrap();
    auto fib_result = std::move(result).unwrap()[0].i32();

    std::cout << "fib(" << n << ") = " << fib_result << " [consumed "
              << consumed << " fuel]\n";
    store.context().set_fuel(kStoreFuel).unwrap();
  }
}
